### PR TITLE
Agentless

### PR DESCRIPTION
**Pull request by hglkrijger**
## Description

Addresses #838

This introduces two new modes of operation for the agent:
- extensions disabled
- agent disabled

### Extensions disabled

This allows an end user to reduce the agent overhead when extensions are not required. This serves customers who have provisioned VMs already and cannot re-provision without the agent. To enable this mode, a new config option `Extensions.Enabled=y|n` is provided, which the user may set. 

When `y` (default), the agent runs without any changes.

When `n`, the goal state interval is reduced to 5 minutes, and extension configuration is not processed. This should help customers who do not want to run the agent, but still need the VM to report status, handle JIT requests, and so on.

### Agent disabled

As of Azure.Compute 2018-06-01, a new option is provided in the ARM API to disable the Linux guest agent completely. Documentation on that is forthcoming. This API will end up setting the `ProvisionGuestAgent` flag in `ovf-env.xml`, which the agent will honor with this change. 

If that flag is set to `false`, the agent will write `/var/lib/waagent/disable_agent`, and if that file exists, sleep forever. No status reports or goal state handling of any kind are processed. 

Once cloud-init supports this flag, the agent service will be disabled and/or removed on provision.

---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/daemon/main.py [hunk 1]
```diff
@@ -99,6 +99,17 @@ def check_pid(self):
 
         fileutil.write_file(pid_file, ustr(os.getpid()))
 
+    def sleep_if_disabled(self):
+        agent_disabled_file_path = conf.get_disable_agent_file_path()
+        if os.path.exists(agent_disabled_file_path):
+            import threading
+            logger.warn("Disabling the guest agent by sleeping forever; "
```
Comment (boumenot) [inline azurelinuxagent/daemon/main.py]: There's no API to disable the agent, there's just a file on disk?

I assumed we would just stop the service, and no code would be running.

File: tests/daemon/test_daemon.py [hunk 2]
```diff
@@ -75,13 +78,58 @@ def test_set_openssl_fips(self, mock_conf, mock_daemon):
...
+
+        self.assertEqual(1, patch_run_provision.call_count)
+        self.assertEqual(1, patch_run_latest.call_count)
+
+    @patch('azurelinuxagent.ga.update.UpdateHandler.run_latest', side_effect=AgentTestCase.fail)
+    @patch('azurelinuxagent.pa.provision.default.ProvisionHandler.run', side_effect=ProvisionHandler.write_agent_disabled)
+    @patch('azurelinuxagent.pa.provision.get_provision_handler', return_value=ProvisionHandler())
+    def test_daemon_agent_disabled(self, _, __, patch_run_latest):
+        """
+        Agent should provision, then sleep forever when disable_agent is found
```
Comment (boumenot) [inline tests/daemon/test_daemon.py]: Why do we start the agent at all?

File: tests/pa/test_provision.py [hunk 3]
```diff
@@ -188,7 +269,48 @@ def test_provision_telemetry_fail(self, mock_util, distro_name,
...
+        self.assertEqual(3, patch_write_agent_disabled.call_count)
+
+        ph.handle_provision_guest_agent(provision_guest_agent='true')
+        self.assertEqual(3, patch_write_agent_disabled.call_count)
+
+        ph.handle_provision_guest_agent(provision_guest_agent='True')
+        self.assertEqual(3, patch_write_agent_disabled.call_count)
+
+        ph.handle_provision_guest_agent(provision_guest_agent='TRUE')
+        self.assertEqual(3, patch_write_agent_disabled.call_count)
```
Comment (boumenot) [inline tests/pa/test_provision.py]: Thorough! 💯

File: README.md [hunk 4]
```diff
@@ -213,6 +214,18 @@ may be used for some string type configuration entries as detailed below.
 
 ### Configuration File Options
 
+#### __Extensions.Enabled__
+
+_Type: Boolean_
```
Comment (boumenot) [inline README.md]: Please add this to the various waagent.conf versions in the tree.  I think people are more likely to read that file than they are the web site.

Comment (hglkrijger) [reply README.md]: Good call, updated

File: azurelinuxagent/daemon/main.py (hunk 1 above)
Comment (hglkrijger) [reply azurelinuxagent/daemon/main.py]: The API is in CRP, exposed through ARM/SDK/CLI. Cloud-init will honor this flag by stopping/disabling the service, but for distros without cloud-init support, I believe this is the best we can do.

File: tests/daemon/test_daemon.py (hunk 2 above)
Comment (hglkrijger) [reply tests/daemon/test_daemon.py]: We do not start it, the system does. If we just exit the service will be restarted. I definitely do not want to get into the business of trying to update the service from a running agent.

**Pull request merged**